virtcontainers/device/drivers/utils.go

```diff
@@ -52,23 +55,17 @@ func isPCIeDevice(bdf string) bool {
 	if len(strings.Split(bdf, ":")) == 2 {
 		bdf = PCIDomain + ":" + bdf
 	}
-	slots, err := ioutil.ReadDir(config.SysBusPciSlotsPath)
+
+	configPath := filepath.Join(config.SysBusPciDevicesPath, bdf, "config")
+	fi, err := os.Stat(configPath)
 	if err != nil {
-		deviceLogger().WithError(err).WithField("path", config.SysBusPciSlotsPath).Warn("failed to list pci slots")
-		return false
-	}
-	b := strings.Split(bdf, ".")[0]
-	for _, slot := range slots {
-		address := getPCISlotProperty(slot.Name(), PCISysFsSlotsAddress)
-		if b == address {
-			maxBusSpeed := getPCISlotProperty(slot.Name(), PCISysFsSlotsMaxBusSpeed)
-			if strings.Contains(maxBusSpeed, PCIeKeyword) {
-				return true
-			}
-		}
+		deviceLogger().WithField("dev-bdf", bdf).WithField("error", err).Warning("Couldn't stat() configuration space file")
+		return false //Who knows?
 	}
-	deviceLogger().WithField("dev-bdf", bdf).Debug("can not find slot for bdf of pci device")
-	return false
+
+	// Plain PCI devices hav 256 bytes of configuration space,
+	// PCI-Express devices have 4096 bytes
+	return fi.Size() > PCIConfigSpaceSize
```

Comment:
I was going to say it might be safer to check if the size is exactly 4096. But presumably if another PCI standard is released that would be a superset of PCI-e (as PCI-e is to PCI) so I guess this is "safe" as-is :smile:

Reply:
Right, and my guess would be that that hypothetical extension will probably behave closer to PCI express than classic PCI.